File: Java/problem1405.java
class Solution {
    private int[] char2num;
    
    private char nextChar(char exclude) {
        char next;
        if (exclude == 'a') {
            next = char2num[1] > char2num[2] ? 'b' : 'c';
        }
        else if (exclude == 'b') {
            next = char2num[0] > char2num[2] ? 'a' : 'c';
        }
        else if (exclude == 'c') {
            next = char2num[0] > char2num[1] ? 'a' : 'b';
        }
        else {
            next = char2num[0] > char2num[1] ? 'a' : 'b';
            next = char2num[next - 'a'] > char2num[2] ? next : 'c';
        }
        return next;
    }
    
    public String longestDiverseString(int a, int b, int c) {
        char2num = new int[] {a, b, c};
        
        char[] result = new char[a + b + c];
        int idx = 0;
        
        while (char2num[0] != 0 || char2num[1] != 0 || char2num[2] != 0) {
            char next;
            if (idx < 2 || result[idx - 1] != result[idx - 2]) {
                next = nextChar(' ');
            }
            else {
                next = nextChar(result[idx - 1]);     
            }
            if (char2num[next - 'a'] <= 0) {
                break;
            }
            --char2num[next - 'a'];
            result[idx++] = next;
        }
        
        return new String(result, 0, idx);
    }
}

File: python/problem1405.py
class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        char2num = {'a' : a,  'b' : b, 'c' : c}

        def nextChar(exclude):
            if exclude == 'a':
                return 'b' if char2num['b'] > char2num['c'] else 'c'
            elif exclude == 'b':
                return 'a' if char2num['a'] > char2num['c'] else 'c'
            elif exclude == 'c':
                return 'a' if char2num['a'] > char2num['b'] else 'b'
            else:
                nc = 'a' if char2num['a'] > char2num['b'] else 'b'
                return nc if char2num[nc] > char2num['c'] else 'c'
        
        result = ''
        while char2num['a'] > 0 or char2num['b'] > 0 or char2num['c'] > 0:
            L = len(result)
            nc = 'a'
            if L < 2 or result[L - 1] != result[L - 2]:
                nc = nextChar(' ')
            else:
                nc = nextChar(result[L - 1])
            if char2num[nc] <= 0:
                break
            char2num[nc] -= 1
            result += nc
        return result
